File: apps/flask/modules/runner.py
from modules.node import Node

class Runner:
	def __init__(self, maze):
		self.open_nodes = []
		self.visited = set()
		self.to_visit = []
		self.start = None
		self.end = None
		self.maze = maze
		self.completed = False
		self.mapped_maze = []
		self.possible_paths = []
		self.path = None
		self.get_open_nodes()
		self.find_end_points()
		
	def get_open_nodes(self):
		p = self.maze.layout
		for x in range(len(p)):
			for y in range(len(p[x])):
				if p[x][y] != self.maze.wall_char:
					self.open_nodes.append(Node((x, y)))
						
	def find_end_points(self):
		def check(node_val):
			if node_val not in [i.value for i in self.open_nodes]:
				node = Node(node_val) 
			else:
				for i in self.open_nodes:
					if i.value == node_val:
						node = i
			return node
			
		for x in range(len(self.maze.layout)):
			for y in range(len(self.maze.layout[x])):
				p = self.maze.layout[x][y]
				if p == self.maze.start_char:
					self.start = check((x, y))
				elif p == self.maze.end_char:
					self.end = check((x, y))

	def look_around(self, node):
		for i in self.open_nodes:
			if i.value[0]-1 == node.value[0] and i.value[1] == node.value[1]:
				node.add_child(i)
			if i.value[0]+1 == node.value[0] and i.value[1] == node.value[1]:
				node.add_child(i)
			if i.value[1]-1 == node.value[1] and i.value[0] == node.value[0]:
				node.add_child(i)
			if i.value[1]+1 == node.value[1] and i.value[0] == node.value[0]:
				node.add_child(i)		
				
	def make_node_paths(self):
		self.to_visit.append(self.start)
		while len(self.to_visit) > 0:
			for point in self.to_visit.copy():
				self.to_visit.remove(point)
				if point not in self.visited:
					self.look_around(point)
					self.visited.add(point)
					new_path = point.path.copy()
					new_path.add(point.value)
					for i in point.children:
						i.set_path(new_path)
						if i.value == self.end.value:
							self.completed = True
						else:
							self.to_visit.append(i)

				
	def view_completed(self):
		x = ""
		for i in self.mapped_maze:
			x += "".join(i)
			x += "\n"
		print(x)
		return x

	def build_path(self, path="x"):
		other_options = set(["x", "o", "+", "*", "p"])
		maze = self.maze	
		if path == maze.start_char or path == maze.end_char or path == maze.wall_char or path == maze.open_char:
			print("Path character is already being used as a maze character trying something else...")
			for i in other_options:
				if i == maze.start_char or i == maze.end_char or i == maze.wall_char or i == maze.open_char:
					pass
				else:
					path = i
					print(f"New path character: {i}")
					break
		self.path = path
		self.mapped_maze = [list(i) for i in maze.layout]
		for i in range(len(self.mapped_maze)):
			for j in range(len(self.mapped_maze[i])):
				if (i, j) in self.end.path and (i, j) != self.start.value:
					self.mapped_maze[i][j] = path

File: apps/next/components/runner.js
import Node from "components/node"

export default (maze) => {
    let r =  {
        maze,
        openNodes: [],
        visited: [],
        toVisit: [],
        start: null,
        end: null,
        completed: false,
        mappedMaze: [],
        possiblePaths: [],
        pathChar: false,
        
        getOpenNodes() {
            let p = this.maze.layout;
            for (let x = 0; x < p.length; x++){
                for (let y = 0; y < p[x].length; y++){
                    if (p[x][y] != this.maze.wallChar) {
                        this.openNodes.push(Node([x, y]));
                    }
                }
            }
        },

        findEndPoints() {
            const check = (nodeVal) => {
                let node = null;
                let onv = this.openNodes.map((i)=> i.value);
                if (!onv.includes(nodeVal)) {
                    node = Node(nodeVal);
                } else {
                    this.openNodes.map((i) => {
                        if (i.value == nodeVal) {
                            node = i;
                        }
                    });
                }
                return node;
            }
            for (let x = 0; x < this.maze.layout.length; x++) {
                for (let y = 0; y < this.maze.layout[x].length; y++) {
                    let p = this.maze.layout[x][y];
                    if (p == this.maze.startChar) {
                        this.start = check([x, y]);
                    } else if (p == this.maze.endChar) {
                        this.end = check([x, y]);
                    }
                }
            }
            return this;
        },

        lookAround(node) {
            this.openNodes.forEach((i) => {
                if (i.value[0] - 1 == node.value[0] && i.value[1] == node.value[1]) {
                    node.addChild(i);
                } else if (i.value[0] + 1 == node.value[0] && i.value[1] == node.value[1]) {
                    node.addChild(i);
                } else if (i.value[1] - 1 == node.value[1] && i.value[0] == node.value[0]) {
                    node.addChild(i);
                } else if (i.value[1] + 1 == node.value[1] && i.value[0] == node.value[0]) {
                    node.addChild(i);
                }
            });
        },

        makeNodePaths() { //error function stuck in forever loop
            this.toVisit.push(this.start);
            while (this.toVisit.length > 0 && this.completed === false) {
                const point = this.toVisit.shift();
                const visitedBool = point.value.length > 0 
                    ? this.visited.map((x)=> x[0] != point.value[0] || x[1] != point.value[1]).every((x)=> x === true)
                    : false;
                if (visitedBool) {
                    this.lookAround(point);
                    this.visited.push(point.value);
                    let newPath = new Set([...point.path]);
                    newPath.add(point.value);
                    point.children.map((i) => {
                        i.setPath(newPath);
                        if (i.value[0] == this.end.value[0] && i.value[1] == this.end.value[1]) {
                            this.end = i
                            this.completed = true;
                        } else {
                            this.toVisit.push(i);
                        }
                    });
                }  
            }
        },

        viewCompleted() {
            if (!this.completed) return console.log("Cannot view completed version of incomplete maze")
            console.log("Completed Maze: ")
            this.mappedMaze.map((i) => console.log(i.join("")));
        },

        buildPath(path="x") {
            if (!this.completed) return console.log("Cannot build path on incomplete maze")
            let otherOptions = ["x", "o", "+", "*", "p"];
            if (path == this.maze.startChar || path == this.maze.endChar || path == this.maze.wallChar || path == this.maze.openChar) { 
                console.log("Path character is already being used as a maze character trying something else...");
                for (let i in otherOptions) { 
                    if (i !== this.maze.startChar && i !== this.maze.endChar && i !== this.maze.wallChar && i !== this.maze.openChar) {
                        path = i;
                        console.log(`New path character: ${i}`);
                        break;             
                    } 
                }
            }
            this.pathChar = path
            this.mappedMaze = this.maze.layout.map((i)=> [...i]);
            for (let i = 0; i < this.mappedMaze.length; i++) {
                for (let j = 0; j < this.mappedMaze[i].length; j++) {
                    [...this.end.path].map((p)=> {
                        if ((p[0] == i && p[1] == j) && (this.start.value[0] != i || this.start.value[1] != j)) {
                            this.mappedMaze[i][j] = path;
                        } 
                    });
                }
            }
        }
    }

    r.getOpenNodes();
    r.findEndPoints();
    return r;
}
